### PR TITLE
Correct stdc++ language and size issues for MSVC

### PR DESCRIPTION
**Pull request by achasveachas**
Description:
This patch makes consistent the type for ssize_t as used on Windows.
We will favor describing it as a ptrdiff_t, which is standard C.
Patches to tclap and nghttp2 are needed for this exercise.

Linux is a 64ILP architecture, so there's no difference between int
and [s]size_t and pointer widths. Windows is a 64P architecture, where
only long long maps to the width of the size_t and C99 standard ptrdiff_t
types.

This leads to a number of places where the Envoy API has presented
an explicit int or int32_t which is smaller than the ptrdiff_t on
Windows. In some places we must cast to or from the API defined width.
Several other misassumptions on the width of int are also addressed.

Windows does not support alternately-named __VA_ARGS__ which is not a
standard C++ language feature in our baseline c++ 17 expectation. Other
minor adjustments reflect other quirks of the MSVC compilier (some more
correct and some simply buggy behavior).

Windows MSVC needs various guards on unrecognized #pragmas, and will not
support any preprocessor operations within the arguments to a macro;
this is not stdc or stdc++.

This patch largely ignores changes which are required for windows, and
do not impact the linux/os-x compilation path. This patch is also
missing the IoHandle abstraction of 'fd' arguments, storage and return
codes, win32 error handling and win32 specific #include's handling.

Signed-off-by: William A Rowe Jr <wrowe@pivotal.io>
Signed-off-by: Yechiel Kalmenson <ykalmenson@pivotal.io>

Risk Level: Moderate
Testing: Passed locally on Windows (with additional patches) and Linux
Docs Changes: None
Release Notes: None

### CODE REVIEW

File: source/common/common/version.cc [hunk 1]
```diff
@@ -19,18 +19,18 @@ const std::string& VersionInfo::revisionStatus() {
 }
 
 const std::string& VersionInfo::version() {
-  CONSTRUCT_ON_FIRST_USE(std::string, fmt::format("{}/{}/{}/{}/{}", revision(),
-                                                  BUILD_VERSION_NUMBER, revisionStatus(),
 #ifdef NDEBUG
-                                                  "RELEASE",
+  const std::string release_type = "RELEASE";
```
Comment (lizan) [inline source/common/common/version.cc]: why this change? make them string_view.

File: source/common/router/header_parser.h [hunk 2]
```diff
@@ -1,3 +1,11 @@
+// grpcpp/grpcpp.h includes <windows.h>, so undef some interfering symbols
```
Comment (lizan) [inline source/common/router/header_parser.h]: ?? this seems a wrong place.

File: test/common/stream_info/test_util.h [hunk 3]
```diff
@@ -16,7 +16,11 @@ class TestStreamInfo : public StreamInfo::StreamInfo {
     memset(&fake_time, 0, sizeof(fake_time));
     fake_time.tm_year = 99; // tm < 1901-12-13 20:45:52 is not valid on macOS
     fake_time.tm_mday = 1;
-    start_time_ = std::chrono::system_clock::from_time_t(timegm(&fake_time));
+#if !defined(WIN32)
+    start_time_ = std::chrono::system_clock::from_time_t(::timegm(&fake_time));
```
Comment (wrowe) [inline test/common/stream_info/test_util.h]: This is a particular problem... There are likely more portable solutions which don't violate the check_validate logic that if failing on the presence of a timegm() call. We simply used the existing logic and would beg another eager volunteer to take on that particular refactoring using stdc++ or absl api, no win32-internals knowledge required.

Comment (lizan) [reply test/common/stream_info/test_util.h]: @jmarantz I think we've been discussing use absl::Time and move away from chrono, will that ultimately solve this issue as well?

Comment (jmarantz) [reply test/common/stream_info/test_util.h]: I would be in favor of moving to absl::Time. However it's a ton of work. I count 1716 std::chrono refs in the codebase.

We'd probably have to do it in a few refactors.

Comment (htuch) [reply test/common/stream_info/test_util.h]: Can't we `sed` this migration? Maybe with some wrappers?

File: bazel/tclap-win64-ull-sizet.patch [hunk 4]
```diff
@@ -0,0 +1,16 @@
+diff --git a/include/tclap/StandardTraits.h b/include/tclap/StandardTraits.h
```
Comment (htuch) [inline bazel/tclap-win64-ull-sizet.patch]: Can you fix this one upstream?

File: bazel/foreign_cc/nghttp2.patch [hunk 5]
```diff
@@ -8,7 +8,7 @@ index 35c77d1d..47bd63f5 100644
    # Set it to "int" to match the behavior of AC_TYPE_SSIZE_T (autotools).
 -  set(ssize_t int)
 +  if(WIN32 AND CMAKE_SIZEOF_VOID_P EQUAL 8)
-+    set(ssize_t int64_t)
++    set(ssize_t ptrdiff_t)
```
Comment (htuch) [inline bazel/foreign_cc/nghttp2.patch]: Can you upstream this? Patches aren't sustainable for us long term.

File: source/common/protobuf/protobuf.h [hunk 6]
```diff
@@ -27,6 +27,14 @@
 #include "google/protobuf/util/type_resolver_util.h"
 #include "google/protobuf/wrappers.pb.h"
 
+#ifdef WIN32
+// one of the above headers includes <windows.h>, so undef some interfering symbols
+#undef TRUE
+#undef DELETE
+#undef ERROR
+#undef GetMessage
```
Comment (htuch) [inline source/common/protobuf/protobuf.h]: This seems pretty fragile, is there some best practice for importing Protobuf headers on Windows or other projects we can crib from?

File: source/common/tracing/http_tracer_impl.h [hunk 7]
```diff
@@ -11,6 +11,13 @@
 #include "common/http/header_map_impl.h"
 #include "common/json/json_loader.h"
 
+#ifdef WIN32
+// <windows.h> defines macros for ERROR and TRUE, so we need to undef them in
+// case it's been included above
+#undef ERROR
+#undef TRUE
```
Comment (htuch) [inline source/common/tracing/http_tracer_impl.h]: Yeah, definitely not a fan of so many seemingly random places that we need to do these `#undef`. Would be very keen to see some way we can hide this and bound the number.

File: source/extensions/filters/network/http_connection_manager/config.cc [hunk 8]
```diff
@@ -170,16 +170,19 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(
                                                                          context_.listenerScope())),
       proxy_100_continue_(config.proxy_100_continue()),
       delayed_close_timeout_(PROTOBUF_GET_MS_OR_DEFAULT(config, delayed_close_timeout, 1000)),
+#ifdef ENVOY_NORMALIZE_PATH_BY_DEFAULT
```
Comment (htuch) [inline source/extensions/filters/network/http_connection_manager/config.cc]: Why this change?

File: test/common/http/http2/codec_impl_test.cc [hunk 9]
```diff
@@ -396,7 +396,7 @@ TEST_P(Http2CodecImplTest, Invalid204WithContentLength) {
   // maximum frame size is 16K. We will add 3,000 headers that will take us above this size and
   // not easily compress with HPACK. (I confirmed this generates 26,468 bytes of header data
   // which should contain a continuation.)
-  for (uint i = 1; i < 3000; i++) {
+  for (unsigned i = 1; i < 3000; i++) {
```
Comment (htuch) [inline test/common/http/http2/codec_impl_test.cc]: Can you just make these explicitly sized types? This is the default style in Envoy.

File: bazel/tclap-win64-ull-sizet.patch (hunk 4 above)
Comment (wrowe) [reply bazel/tclap-win64-ull-sizet.patch]: We will submit it, but it is needed locally short term.

File: test/common/http/http2/codec_impl_test.cc (hunk 9 above)
Comment (wrowe) [reply test/common/http/http2/codec_impl_test.cc]: That's fine IFF these are envoy args passed to envoy functions alone.

It's a horrible style if consuming stdcxx/absl classes, your intwidth assumptions are either wrong today (Win32 is 32, Unix is 32 or 64) or they become wrong tomorrow (128LP architectures?)

Two things are going on. MSVC is especially picky about confusing unsigned for signed types. And reflectors and emplacement types must match identically. So if the int is going to a stdcxx/absl class, using fixed width types is very ill-advised.

This one could be either 'unsigned int' or 'uint32_t', perhaps. It's only uint that isn't a C[++] type at all.

File: test/common/stream_info/test_util.h (hunk 3 above)
Comment (wrowe) [reply test/common/stream_info/test_util.h]: Is this item worth it's own issue, maybe? I'd beg for this bit to be accepted in the short-term.

File: bazel/foreign_cc/nghttp2.patch (hunk 5 above)
Comment (wrowe) [reply bazel/foreign_cc/nghttp2.patch]: This will not be accepted upstream until a version major bump (note it is our patch - of envoy's present patch.) PR and discussion item here; https://github.com/nghttp2/nghttp2/pull/1395

File: source/common/protobuf/protobuf.h (hunk 6 above)
Comment (wrowe) [reply source/common/protobuf/protobuf.h]: I'll attach the explanation top-post in the review comments, but I agree we should consume our own platform.h here and delegate inclusion of windows.h at that inflection point, trashing MS macros at the time we include windows.h. Individual modules should never be injecting MS SDK API headers anywhere without coordination.

**Comment by wrowe** [comment]
As a top-level comment, feel free to push back any uio.h vs. winsock2.h (and undefines.) The next submission this coming week makes a lot of changes, to localize all of those inclusions in platform.h as our single point of contact with the platform-specific networking API.

File: bazel/tclap-win64-ull-sizet.patch (hunk 4 above)
Comment (htuch) [reply bazel/tclap-win64-ull-sizet.patch]: It seems small enough that we could kick this into the pipeline in parallel. We really dislike patches, they are only a solution when all else fails when working with an upstream project.

File: test/common/stream_info/test_util.h (hunk 3 above)
Comment (htuch) [reply test/common/stream_info/test_util.h]: Yes, it's definitely a distinct issue. Can you open a GH issue and tag this as TODO?

File: source/common/tracing/http_tracer_impl.h (hunk 7 above)
Comment (htuch) [reply source/common/tracing/http_tracer_impl.h]: Fantastic. In this case, I would suggest you formalize this rule in `tools/header_order.py`, so that we have a machine enforcement of this logic.

File: source/common/protobuf/protobuf.h (hunk 6 above)
Comment (htuch) [reply source/common/protobuf/protobuf.h]: Thanks, I like this solution.

File: source/common/router/header_parser.h (hunk 2 above)
Comment (lizan) [reply source/common/router/header_parser.h]: I mean this could never go before #pragma once, and grpcpp is not included in this file either.

**Comment by wrowe** [comment]
Simply to unstale, the two action items I'm pursuing are the request to string_view as identified by lizan, and either retract or advance the use of platform.h for <windows.h> (winsock2 etc) inclusion, and finally open a ticket to refactor to absl time classes that wouldn't fall under this PR. If I have missed anything else that prevents this patch from being pushed, please identify such.

File: source/common/common/version.cc (hunk 1 above)
Comment (wrowe) [reply source/common/common/version.cc]: This is another example where the macro CONSTRUCT_ON_FIRST_USE cannot have nested #cpp statements such as #ifdef.  We simply followed the file's current convention, if a rewrite is in order, doesn't it make sense to convert this all to absl::string_view? Is it possible to accept this patch as-is, to accomplish the stated purpose of this PR (to get it compiling without gcc'isms)?

File: bazel/tclap-win64-ull-sizet.patch (hunk 4 above)
Comment (wrowe) [reply bazel/tclap-win64-ull-sizet.patch]: If and when we ever see a tclap 1.4 release, this patch and the defect go away, since the entire issue was refactored away 6 years ago;
https://sourceforge.net/p/tclap/code/ci/5d4ffbf2db794af799b8c5727fb6c65c079195ac/
Please consider this patch acceptable short-term (or permanently given the stable/no-releases state) for this component and we'll treat this as resolved for modern windows 64 compilers.

File: source/common/common/version.cc (hunk 1 above)
Comment (lizan) [reply source/common/common/version.cc]: What's the gcc'ism in this function before your change? You should be able to make L23-L30 to `const absl::string_view`, basically no string allocation other than the last statement.

File: source/common/router/header_parser.h (hunk 2 above)
Comment (wrowe) [reply source/common/router/header_parser.h]: Agreed. Tested and removing.

File: test/common/http/http2/codec_impl_test.cc (hunk 9 above)
Comment (wrowe) [reply test/common/http/http2/codec_impl_test.cc]: Just to close this discussion, in all of these cases, what is 32 bits on windows and 64 bits on linux can't be used as an argument to a cb function declaration or emplace_back invocation. These types are each correct, tested and re-tested on both linux and windows for correctness according to the class declarations.

File: source/common/common/version.cc (hunk 1 above)
Comment (achasveachas) [reply source/common/common/version.cc]: @lizan we updated the file, is this what you had in mind?

File: bazel/tclap-win64-ull-sizet.patch (hunk 4 above)
Comment (htuch) [reply bazel/tclap-win64-ull-sizet.patch]: Ack, please leave a reference to this thread in the `patches` arg of the `_repository_impl` .

File: source/extensions/filters/network/http_connection_manager/config.cc (hunk 8 above)
Comment (htuch) [reply source/extensions/filters/network/http_connection_manager/config.cc]: Can you share a citation to the standard for my benefit? Up until this point, I was under the impression that the preprocessor was a purely lexical step that occurred prior to any parsing or consideration of the language syntax, but I prefer to consume my standards lazily, they don't make great beach reading :)

File: test/common/http/http2/codec_impl_test.cc (hunk 9 above)
Comment (htuch) [reply test/common/http/http2/codec_impl_test.cc]: OK, but please consider that Envoy style remains fixed types, so where we're not forced to do this for interop with STL etc. let's keep things as they are. I haven't spotted any violations of this in this PR, but worth keeping in mind. Resolved.

File: test/integration/echo_integration_test.cc [hunk 10]
```diff
@@ -49,15 +49,16 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, EchoIntegrationTest,
 TEST_P(EchoIntegrationTest, Hello) {
   Buffer::OwnedImpl buffer("hello");
   std::string response;
-  RawConnectionDriver connection(
+  std::unique_ptr<RawConnectionDriver> connection;
```
Comment (htuch) [inline test/integration/echo_integration_test.cc]: Can you elaborate on why this change  was needed?

File: source/extensions/filters/network/http_connection_manager/config.cc (hunk 8 above)
Comment (wrowe) [reply source/extensions/filters/network/http_connection_manager/config.cc]: [Found the citation;  ISO/IEC 9899:2011 6.10.3 
```
The sequence of preprocessing tokens bounded by the outside-most matching parentheses forms the list of arguments for the function-like macro. The individual arguments within the list are separated by comma preprocessing tokens, but comma preprocessing tokens between matching inner parentheses do not separate arguments. If there are sequences of preprocessing tokens within the list of arguments that would otherwise act as preprocessing directives, the behavior is undefined.
```
The C Preprocessor is a serial line-oriented grammar, with one well understood line break exception sequence of '\' <newline> (which is evaluated as a single space.)
```
#define sum_values(arg1, arg2) ( (arg1 + arg2) )
int next(int value)
{
    /* The C preprocessor is expected to replace comments with a space
     * and smash whitespace while performing the macro expansion, so...
     */
    return sum_values(value,
#ifdef TWO_IF_BY_SEA
                      2
#else
                      1
#endif
    );
}
```
On some compilers, the #if expression is evaluated, in spite of interrupting macro evaluation and expansion.

On other compilers, the macro invocation is evaluated as;
```
    return ( (value) + (#ifdef TWO_IF_BY_SEA 2 #else 1 #endif) );
```
I was able to track down the details of this specific issue with some help from CMU's SEI CERT C Coding Standard, lots more great beach reading for you :)
https://wiki.sei.cmu.edu/confluence/display/c/CC.+Undefined+Behavior#CC.UndefinedBehavior-ub_93

File: source/common/router/retry_state_impl.cc [hunk 11]
```diff
@@ -92,7 +92,7 @@ RetryStateImpl::RetryStateImpl(const RetryPolicy& route_policy, Http::HeaderMap&
 
   if (!retriable_request_headers_.empty()) {
     // If this route limits retries by request headers, make sure there is a match.
-    bool request_header_match = false;
+    uint32_t request_header_match = false;
```
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/router/retry_state_impl.cc]: It isn't used as a bool. It's used for boolean math against a uint32_t.

File: test/integration/echo_integration_test.cc (hunk 10 above)
Comment (wrowe) [reply test/integration/echo_integration_test.cc]: I have literally no clue, it was before my time, and not suggesting this is the single best workaround.  Reverting the patch on msvc we end up with;
```
ERROR: C:/workspace/envoy/test/integration/BUILD:565:1: C++ compilation of rule '//test/integration:echo_integration_test_lib_internal_only' failed (Exit 2)
test/integration/echo_integration_test.cc(56): error C2065: 'connection': undeclared identifier
test/integration/echo_integration_test.cc(106): error C2065: 'connection': undeclared identifier
Target //test/integration:echo_integration_test failed to build
```
We are happy to withdraw changes to this specific file for separate consideration.

File: bazel/tclap-win64-ull-sizet.patch (hunk 4 above)
Comment (wrowe) [reply bazel/tclap-win64-ull-sizet.patch]: Added short note + refactor link

File: test/integration/echo_integration_test.cc (hunk 10 above)
Comment (wrowe) [reply test/integration/echo_integration_test.cc]: We see an additional problem on Windows with this test. We will bypass it short term and are withdrawing the changes to echo_integration_test until they are better understood by all.

File: source/common/tracing/http_tracer_impl.h (hunk 7 above)
Comment (wrowe) [reply source/common/tracing/http_tracer_impl.h]: Please review our submission of yesterday, localizing the decisions on posix / windows API headers and fix-ups of these to the include/envoy/common/platform.h file.

Expanding on that idea, it would also be possible to resolve the problems with Linux <sys/types.h> and #undef the 'major', 'minor' defines which cause significant noise at certain points in the build. It wouldn't be unreasonable to move this (basically portable) inclusion into platform.h as well.

**Comment by achasveachas** [comment]
Right now the only formatting failures are the ones associated with the `gmtime` fix, would you rather we pull that out until that refactor (#8717 ) is addressed so the Linux CI can run properly?

File: source/common/common/fmt.h [hunk 12]
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include "envoy/common/platform.h" // Avert format.h including windows.h
```
Comment (htuch) [inline source/common/common/fmt.h]: Can we fix `fmtlib` upstream?

File: source/common/network/cidr_range.cc [hunk 13]
```diff
@@ -1,8 +1,10 @@
 #include "common/network/cidr_range.h"
 
+#ifndef WIN32
```
Comment (htuch) [inline source/common/network/cidr_range.cc]: Why not platform.h?

File: source/common/router/retry_state_impl.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (htuch) [reply source/common/router/retry_state_impl.cc]: Can we change the `request_header_match = true` to `request_header_match = 1`? I think this is clearer if we switch the type here.

File: source/extensions/filters/network/http_connection_manager/config.cc (hunk 8 above)
Comment (htuch) [reply source/extensions/filters/network/http_connection_manager/config.cc]: Ack, thanks for the explanation.

File: source/common/common/fmt.h (hunk 12 above)
Comment (wrowe) [reply source/common/common/fmt.h]: As a practical matter? Any of our 3 dozen dependencies could introduce an #include <window.h> and disturb our source compilation, that header is not meant for general C++ language consumption. By this, we mean that MS heavily leverages #define and C++ heavily leverages multiple legible tokens. The logic in C language might introduce add_interface() but in C++ we are going to introduce interface.add(), which collides with windows.h defines.

We've introduced platform.h to work around this every time we encounter the issue. And if it is fixed upstream, then we could drop our #include, but it isn't critical. The answer at envoy is to eventually stop using Posix or Windows types in our own .h interface definitions, and localize the Posix and Windows implementations to our .cc sources (which will always include platform.h.) Exceptions would include our .h interface stubs to third party libraries such as grpc, nghttp2, etc.

I don't believe an upstream fix is really a priority, since I'm reading on #envoy-dev that fmtlib is being phased out for absl? It certainly shouldn't block this patch.

File: source/common/router/retry_state_impl.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/router/retry_state_impl.cc]: Switched these two constants at 95 and 98.

File: source/common/network/cidr_range.cc (hunk 13 above)
Comment (wrowe) [reply source/common/network/cidr_range.cc]: An oversight. Corrected, thanks!

File: source/common/common/fmt.h (hunk 12 above)
Comment (htuch) [reply source/common/common/fmt.h]: I feel that any 1st class support for Windows in Envoy doesn't end at Envoy. With my maintainer hat on I'm labeling this technical debt.. It's just a single line, but one day someone will come along and spend a day or so chasing down some obscure issue that is masked by this.

At the same time, we're pragmatic about tech debt and want to see Windows support land.

I would like to see tracking issues filed with the respective communities or in Envoy GH and referenced in these comments. Even if you folks aren't going to own this, we need to recognize these problems and provide the opportunity for other folks to do so.

Comment (htuch) [reply source/common/common/fmt.h]: Also, we will live with `fmtlib` for a long time. It's deprecated on the data path, but it's a key part of logging which won't be going away.

**Comment by wrowe** [comment]
> LGTM modulo filing requested tracking issues and updating comments. This is epic, thanks for the major cleanup!

Some confusion, I'm wondering this is modulo what remaining items? All tracking issues except your fmt class suggestion were filed upstream that I'm aware of, all requested comments had been added in this patch that I'm aware of.

So if there is anything to do other than pushing an upstream a suggestion to the fmt project to hide the inclusion of the windows.h header from all general consumers of their headers (a very good suggestion to push), I've missed it, other than @lizan 's final suggestion of 'static const'ness. Please point out any other remaining items, thanks!

File: source/common/common/version.cc [hunk 14]
```diff
@@ -19,18 +21,18 @@ const std::string& VersionInfo::revisionStatus() {
 }
 
 const std::string& VersionInfo::version() {
-  CONSTRUCT_ON_FIRST_USE(std::string, fmt::format("{}/{}/{}/{}/{}", revision(),
-                                                  BUILD_VERSION_NUMBER, revisionStatus(),
 #ifdef NDEBUG
-                                                  "RELEASE",
+  const absl::string_view release_type = "RELEASE";
```
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/common/version.cc]: Resolved and pushed. Thanks!

File: source/common/common/fmt.h (hunk 12 above)
Comment (wrowe) [reply source/common/common/fmt.h]: > Also, we will live with fmtlib for a long time. It's deprecated on the data path, but it's a key part of logging which won't be going away.

Fair point, and researching, it turns out they have already responded.

https://github.com/fmtlib/fmt/issues/171

We have the option of disabling fmt.h's  use of ``windows.h``, we just need to define ``FMT_USE_WINDOWS_H=0`` as appropriate. This simply deprives us of the optimizations present in bazel-envoy/external/com_github_fmtlib_fmt/include/fmt/format-inl.h

It's an open question whether we want to define this symbol prior to including format.h in our 
source/common/common/fmt.h header or include windows.h before they can, these have the same effect. If we want to explore this further, perhaps an issue is in order for further discussions, but we are reasonably happy with the proposed solution.

**Comment by wrowe** [comment]
> [...] modulo filing requested tracking issues and updating comments.

We believe these are all filed, comments updated, code adjusted as requested.

**Comment by htuch** [comment]
Thanks @wrowe, ready to ship when CI clears.

**Comment by kfaseela** [comment]
There is a tclap 1.2.5 version available, will this patch cause any trouble if we try to upgrade to 1.2.5?
